src/community/voronoi.c

```diff
@@ -0,0 +1,638 @@
+
+        newx = peakx(x1, x2, x3, f1, f2, f3);
+        IGRAPH_CHECK(f(newx, &newf, extra));
+
+        /* We need to decide whether we drop (x1, f1) or (x2, f2) for the following iterations.
+         * The sign of a1 (or a2) determines whether dropping x1 (or x2) yields a convex or concave
+         * parabola in the next iteration. We need a negative sign = concave parabola.
+         * We always keep (x3, f3) as it was the last added point. */
+        igraph_real_t a1 = coeff2(x2, x3, newx, f2, f3, newf);
+        igraph_real_t a2 = coeff2(x1, x3, newx, f1, f3, newf);
+
+        /* We cannot continue without the Brent optimizer switching to minimization.
+         * Terminate search, accepting the current result. */
+        if (a1 >= 0 && a2 >= 0) {
+            break;
+        }
+
+        if (a1 <= a2) {
+            x1 = x2;
+            x2 = x3;
+            x3 = newx;
+
+            f1 = f2;
+            f2 = f3;
+            f3 = newf;
+        } else {
+            x2 = x1;
+            x1 = x3;
+            x3 = newx;
+
+            f2 = f1;
+            f1 = f3;
+            f3 = newf;
+        }
+
+        /* Check if value goes out of initial interval. */
+        if (x3 < lo || x3 > hi) {
+            IGRAPH_ERROR("Optimizer did not converge while maximizing modularity for Voronoi communities.",
+                         IGRAPH_DIVERGED);
+        }
+
+        /* We exploit the fact that we are optimizing a discrete valued function, and we can
+         * detect convergence by checking that the function value stays exactly the same.
+         *
+         * As an optimization, we only check whether the two of the three f values are the same.
+         * Almost always, when this is the case, another iteration would not yield a better
+         * maximum, however, saving a call to f() improves performance noticeably.
+         */
+        const igraph_real_t eps = 1e-10;
+        int c1 = igraph_cmp_epsilon(f1, f3, eps);
+        int c2 = igraph_cmp_epsilon(f2, f3, eps);
+        if (c1 == 0 || c2 == 0) {
+            break;
+        }
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/* Work data for get_modularity() */
+typedef struct {
+    const igraph_t *graph;
+    const igraph_vector_t *local_dens;
+    const igraph_vector_t *lengths;
+    const igraph_vector_t *weights;
+    igraph_neimode_t mode;
+    igraph_vector_int_t *generators;
+    igraph_vector_int_t *membership;
+    igraph_real_t modularity;
+} get_modularity_work_t;
+
+
+/* Objective function used with brent_opt(), it computes the modularity for a given radius. */
+static igraph_error_t get_modularity(igraph_real_t r, igraph_real_t *modularity, void *extra) {
+    get_modularity_work_t *gm = extra;
+
+    IGRAPH_CHECK(choose_generators(gm->graph, gm->generators, NULL,
+                                   gm->local_dens, gm->lengths, gm->mode,
+                                   r));
+    igraph_voronoi(gm->graph, gm->membership, NULL,
```

Comment:
Why don't we need an `IGRAPH_CHECK(...)` around this? (and later in the call to `igraph_modularity()`)?

Reply:
Oops ... will fix it.